test/image_asset_test.cpp

```diff
@@ -0,0 +1,112 @@
+#include <rive/core/binary_reader.hpp>
+#include <rive/file.hpp>
+#include <rive/node.hpp>
+#include <rive/shapes/clipping_shape.hpp>
+#include <rive/shapes/rectangle.hpp>
+#include <rive/shapes/image.hpp>
+#include <rive/assets/image_asset.hpp>
+#include <rive/relative_local_asset_resolver.hpp>
+#include "no_op_renderer.hpp"
+#include <catch.hpp>
+#include <cstdio>
+
+TEST_CASE("image assets loads correctly", "[assets]")
+{
+	FILE* fp = fopen("../../test/assets/walle.riv", "r");
+	REQUIRE(fp != nullptr);
+
+	fseek(fp, 0, SEEK_END);
+	auto length = ftell(fp);
+	fseek(fp, 0, SEEK_SET);
+	uint8_t* bytes = new uint8_t[length];
+	REQUIRE(fread(bytes, 1, length, fp) == length);
+	auto reader = rive::BinaryReader(bytes, length);
+	rive::File* file = nullptr;
+	auto result = rive::File::import(reader, &file);
+
+	REQUIRE(result == rive::ImportResult::success);
+	REQUIRE(file != nullptr);
+	REQUIRE(file->artboard() != nullptr);
+
+	auto node = file->artboard()->find("walle");
+	REQUIRE(node != nullptr);
+	REQUIRE(node->is<rive::Image>());
+	auto walle = node->as<rive::Image>();
+	REQUIRE(walle->imageAsset() != nullptr);
+	REQUIRE(walle->imageAsset()->decodedByteSize == 218873);
+
+	auto eve_left = file->artboard()->find("eve_left");
+	REQUIRE(eve_left != nullptr);
+	REQUIRE(eve_left->is<rive::Image>());
+	REQUIRE(eve_left->as<rive::Image>()->imageAsset() != nullptr);
+	REQUIRE(eve_left->as<rive::Image>()->imageAsset()->decodedByteSize ==
+	        246825);
+
+	auto eve_right = file->artboard()->find("eve_right");
+	REQUIRE(eve_right != nullptr);
+	REQUIRE(eve_right->is<rive::Image>());
+	REQUIRE(eve_right->as<rive::Image>()->imageAsset() != nullptr);
+	REQUIRE(eve_right->as<rive::Image>()->imageAsset() != walle->imageAsset());
+	REQUIRE(eve_right->as<rive::Image>()->imageAsset() ==
+	        eve_left->as<rive::Image>()->imageAsset());
+
+	file->artboard()->updateComponents();
+
+	rive::NoOpRenderer renderer;
+	file->artboard()->draw(&renderer);
+
+	delete file;
+	delete[] bytes;
+}
+
+TEST_CASE("out of band image assets loads correctly", "[assets]")
+{
+	std::string filename = "../../test/assets/out_of_band/walle.riv";
+	rive::RelativeLocalAssetResolver resolver(filename);
```

Comment:
Notice how the asset resolver needs to be specifically instanced for a filename. This local resolving implementation requires the filename as it uses it to determine the relative path to the asset.

Reply:
Exactly, it's just a specific implementation. The abstraction is exactly what you're talking about, only maybe even more abstract, it just says, "I need the matching bytes for this FileAsset". The local resolver uses that to build up a path based on the asset name and id. Which is what the exporter uses for the unique filename, you'll see that's in FIleAsset:
https://github.com/rive-app/rive-cpp/blob/8aa6f797f18a7a0075ba816ef827b7c556f23816/src/assets/file_asset.cpp#L20-L32

But the general abstraction is just this, you could load them over the network, from a bundle/zip, whatever!
https://github.com/rive-app/rive-cpp/blob/8aa6f797f18a7a0075ba816ef827b7c556f23816/include/rive/file_asset_resolver.hpp#L17